pyiceberg/partitioning.py

```diff
@@ -272,6 +272,60 @@ def assign_fresh_partition_spec_ids(spec: PartitionSpec, old_schema: Schema, fre
 T = TypeVar("T")
 
 
+class PartitionMap(Generic[T]):
+    _specs: dict[int, PartitionSpec]
+    _partition_maps: dict[int, dict[Record | None, T]]
+
+    def __init__(self, specs: dict[int, PartitionSpec]):
+        self._specs = specs
+        self._partition_maps = {}
+
+    def __len__(self) -> int:
+        """Return the length of the partition map.
+
+        Returns:
+            length of _partition_maps
+        """
+        return len(self.values())
+
+    def is_empty(self) -> bool:
+        return len(self._partition_maps.values()) == 0
+
+    def contains_key(self, spec_id: int, struct: Record) -> bool:
+        return self._partition_maps.get(spec_id) is not None
```

Comment:
Should we check if the struct key is present?

```suggestion
    def contains_key(self, spec_id: int, struct: Record) -> bool:
        return struct in self._partition_maps.get(spec_id, {})
```